presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/joins/JoinGraph.java

```diff
@@ -200,15 +218,28 @@ private JoinGraph joinWith(JoinGraph other, List<JoinNode.EquiJoinClause> joinCl
     private static class Builder
             extends PlanVisitor<Context, JoinGraph>
     {
+        // TODO When com.facebook.presto.sql.planner.optimizations.EliminateCrossJoins is removed, remove 'shallow' flag
```

Comment:
using Visitors with a hacked Lookup is acceptable to me when that Visitor is calculating the trait kind of value (like is scalar or is sorted etc).

Here I think we need a different approach. I would first create a MultiJoinNode and then create a graph from it and then replace (inline) it with JoinNodes.